File: 2020/18_operation_order/aoc18.jl
# Operation Order
#
# Advent of Code 2020, day 18
# Solution by Geir Arne Hjelle, 2020-12-18
module AOC
using Pipe

function add_to_value(value, op, number)
    if op == '+'
        value + parse(Int, number)
    elseif op == '*'
        value * parse(Int, number)
    end
end

function parse_left_right(expression)
    value, op, number = 0, '+', ""
    for char in expression
        if char ∈ ['+', '*']
            value = add_to_value(value, op, number)
            number = ""
            op = char
        elseif char ∈ "0123456789"
            number = number * char
        end
    end
    add_to_value(value, op, number)
end

function parse_plusses_first(expression)
    while true
        m = match(r"((\d+) [+] (\d+))", expression)
        if m === nothing
            break
        end
        value = parse(Int, m[2]) + parse(Int, m[3])
        expression = replace(expression, m[1] => string(value), count=1)
    end
    parse_left_right(expression)
end

function parse_parens(calculate)
    function _parse_parens(line)
        while true
            # Match unnested parentheses
            m = match(r"(\([^)(]*?\))", line)
            if m === nothing
                break
            end
            value = calculate(m[1][2:end - 1])
            line = replace(line, m[1] => string(value))
        end
        calculate(line)
    end
    _parse_parens
end

# Solve the problem for one file
function solve(filename)
    println("\n$(filename)")

    # Read from file
    input = open(filename) do fid
        fid |> readlines
    end

    # Part 1
    input .|> parse_parens(parse_left_right) |> sum |> println

    # Part 2
    input .|> parse_parens(parse_plusses_first) |> sum |> println

end


# Solve the problem for each file
ARGS .|> solve
end

File: 2020/18_operation_order/aoc18.py
"""Operation Order

Advent of Code 2020, day 18
Solution by Geir Arne Hjelle, 2020-12-18
"""
import pathlib
import re
import sys

debug = print if "--debug" in sys.argv else lambda *_: None


class EvaluatingInt(int):
    """Integers that can evaluate expressions"""

    @classmethod
    def evaluate(cls, expression):
        """Evaluate expression"""

        def wrap_class(match):
            """Wrap a regex match in a class constructor"""
            return f"{cls.__name__}({match.group()})"

        return eval(re.sub(r"\d+", wrap_class, expression.translate(cls.ops)))


class LeftRightInt(EvaluatingInt):
    """Integers that add and multiply left to right

    Use - for multiplication, so addition and multiplication have the same
    precedence
    """

    ops = str.maketrans({"+": "+", "*": "-"})

    def __add__(self, other):
        return self.__class__(super().__add__(other))

    def __sub__(self, other):
        return self.__class__(super().__mul__(other))


class PlusFirstInt(EvaluatingInt):
    """Integers where addition have precedence over multiplication

    Use * for addition and + for multiplication to flip precedence
    """

    ops = str.maketrans({"+": "*", "*": "+"})

    def __mul__(self, other):
        return self.__class__(super().__add__(other))

    def __add__(self, other):
        return self.__class__(super().__mul__(other))


def main(args):
    """Solve the problem for all file paths"""
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        solve(file_path)


def solve(file_path):
    """Solve the problem for one file path"""
    print(f"\n{file_path}:")
    expressions = file_path.read_text().strip().split("\n")

    # Part 1
    results = [LeftRightInt.evaluate(e) for e in expressions]
    print(f"The total sum is {sum(results)}")

    # Part 2
    results = [PlusFirstInt.evaluate(e) for e in expressions]
    print(f"The total sum is {sum(results)}")


if __name__ == "__main__":
    main(sys.argv[1:])
